Mesh diagnostics in the mesh generation phase

GiudGiud opened the pull request:
Can be used at any point.

Bunch of pre-selected diagnostics, still missing:
- sideset orientation.     DONE
- waterproofness of sidesets (tough one).  WILL NOT DO
- negative Jacobians.     DONE

I just did a 1M basic mesh and the whole diagnostics completed in seconds.
Of course it was not filled with complex non-conformalities but it s still boding very well for how usable this will be for serialized meshes

one comment is that while this appropriately runnable at any point during the mesh generator, it s not necessarily my preferred structure for this.
I would like some sort of Diagnostics system trigger-able at any point in time, in Mesh generation or in the simulation, and a lot of the FEProblemBase integrity diagnostics would go there too. 
This is more of a longer term unfunded task though, so saving it for later. But the added value & lower maintenance cost would be significant imo.

Another comment is that most of the existing diagnostics rely on ThreadedXXXLoop instead of what we have here, which are a bunch of unthreaded loops. I think since the runtimes are so slow that's ok. But it goes to show we don't quite follow the old paradigm with this MeshDiagnosticsGenerator

Commit:
add MeshDian=gnosticGenerator that can
examine the elemen types in a subdomain
check the min and max sizes of the element's volume
examine volume of the elements

Commit:
Testing the functions of the MeshDiagnosticsGenerator:
Element types in a subdomain
Volume sizes of elements (mins and maxs)

Commit:
add check for nonplanar sides

Commit:
Add diagnostic for non-conformality

Commit:
Add tests for non-conformality

Commit:
Add test for non plananr sides

Commit:
Add test for node-based heuristic on element overlap

Commit:
WIP examine non conformalities

Commit:
Saving before making changes from previous files

Commit:
Limit diagnostics verbosity

GiudGiud committed:
Re-organize testing folder
Add tests specifications

more work on diagnostics test

GiudGiud committed:
Const-ness of mesh diagnostics generator atrributes

GiudGiud committed:
Cleaner logging of diagnostics results on a per-diagnostic level

GiudGiud committed:
Strengthen tests by using regex

GiudGiud committed:
More logging, more comments on why this is for replicated meshes

GiudGiud committed:
Handle elements with 3 consecutive aligned nodes for non-planity check
refs #25278

GiudGiud committed:
Add AMR-driven non-conformality detection for quad elements
refs #25280

GiudGiud committed:
Add AMR-driven non-conformality detection for tri elements
refs #25280

GiudGiud committed:
Better heuristic to handle two levels of refinement across a single interface

GiudGiud committed:
Add support for TET refinement detection

GiudGiud committed:
Add tests for AMR detection
- tri & quads
- basic, second order, two levels of refinement across the interface

GiudGiud committed:
Add several accomodations for tet refinement detection:
- allow not all children of the tentative parent to match
- allow for several coarse elements found near the refined ones (refinement happens at an edge)
- second order treatment: check for vertex status of nodes
- new heuristics for finding potentially refined tets

GiudGiud committed:
Add test for tet refinement detection
- basic & second order
- no two levels of refinement across as neighbors arent found well in that case

GiudGiud committed:
Add support for HEX refinement detection
_ factor out node-ordering between hex & quad needs
refs #25280

GiudGiud committed:
Add test for hex refinement detection
- basic, second order, two levels across

GiudGiud committed:
Clean up debugging statements
refs #25278

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1718542/) on 8b1106b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25087/site)

This comment will be updated on new commits.

GiudGiud committed:
Add a check on sidesets being oriented consistently between subdomains
refs #25278

GiudGiud committed:
Add test for inconsistently oriented sidesets

GiudGiud committed:
Remove uninitialized memory issues

GiudGiud committed:
Add a test for no-checks on
and a simple input to turn on all diagnosis at once

GiudGiud committed:
Add diagnostic for negative Jacobians

GiudGiud committed:
Add documentation for all diagnostics

GiudGiud committed:
Add test for negative Jacobian:
make a butterfly quad. It will have a negative Jacobian somewhere
Too distorted though and you hit other failures

GiudGiud committed:
Fuse two loops for hex / elem amr detection
Better logs for negative Jacobian: we're counting elements not Qps

GiudGiud committed:
More constness in local variables
Remove class attribute counters, not needed

GiudGiud committed:
Civet failures
- set cant hold const pair for some reasons
- dont leave variables uninitialized
- fix sqa tests items
- dont de-allocate and de-allocate mesh refiner all the time
- fix logic for hex/quads
- fix parameter links in doco

loganharbour commented:
> set cant hold const pair for some reasons

this is the case for all STL containers; things need to be copy assignable and thus can't have const members

loganharbour commented on framework/src/utils/Conversion.C:
```diff
@@ -517,6 +517,92 @@ stringify(ElementType t)
   }
 }
 
+std::string
```
https://mooseframework.inl.gov/docs/doxygen/libmesh/namespacelibMesh_1_1Utility.html#aa99a42a5efab878469e7f58e5c033335

with `libMesh::enum_to_strong<ElemType>(t)`

GiudGiud replied:
this is done already in devel. Just need a next merge to rebase and use that code

GiudGiud replied:
done

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1730325/) on 57e9ca2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5473cf</th>
      <th colspan=4>#25087 57e9ca</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.70%</td>
      <td>+0.04%</td>
      <td>92.28%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90278</td>
      <td>90865</td>
      <td>+587</td>
      <td>586</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15117</td>
      <td>15165</td>
      <td>+48</td>
      <td>49</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25087/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25087/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25087/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Sorry for the big PR. This PR being a single object with distinct sections for each diagnostic I think review complexity remains low, while review size is indeed high.

lindsayad commented:
@roystgnr maybe you should lead the review on this one?

roystgnr commented on framework/doc/content/source/meshgenerators/MeshDiagnosticsGenerator.md:
```diff
@@ -0,0 +1,136 @@
+!alert note
+A Lagrange finite element family of the default order for each element is used for this check, along with a fourth-order
+Gauss quadrature. If you intend to use another finite element family,
+feel free to modify the generator for your needs.
```
Ah, I get it now, after looking through the code.  You're using `FE::reinit()` rather than just creating an `FEMap` and `QGauss` and calling the methods at that level by hand, and you have to pick *some* type of FE.  Our fault at the libMesh level for not making the lower-level methods more convenient to use; your code looks fine given that context.

But it doesn't actually matter that your FE type is `LAGRANGE`.  It'd be more efficient to make it a `CONSTANT` `MONOMIAL`, in fact.  And I definitely wouldn't talk about the FE family in the docs here; that's a trick that doesn't affect the results, unlike the quadrature rule.

GiudGiud replied:
ah I should have known. I started out realizing the quadrature mattered but never thought again about the FE family

roystgnr commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+
+      // Use the default element order
+      // We would error if going above what the element can't support
+      const FEType fe_type(elem->default_order(), libMesh::LAGRANGE);
```
I'd just create a constant `MONOMIAL` here.  The map is the map regardless, and if we're going to have to have shape functions because that's the easiest API to use, we should at least evaluate the cheapest shape functions we can ... and for that matter, we shouldn't evaluate them - prerequest `get_JxW` to tell the `FE` that it needs to evaluate the map, and then it won't fall back on the (deprecated!) "I got no prerequests, best evaluate everything to be safe" code path.

GiudGiud replied:
changed

roystgnr commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+    for (const auto & elem : mesh->element_ptr_range())
+    {
+      // Get a high-ish order quadrature
+      QGauss qrule(mesh->mesh_dimension(), FOURTH);
```
It wouldn't be crazy to use `FIFTH` order or user-selectable order or even loop over quadrature order here.  In a real simulation, even a quadratic FE type is likely to trigger a `FIFTH` order quadrature, so with `FOURTH` we're already potentially missing barely-singular elements that could fail in practice.  And although I know MOOSE users tend to stay low order, IIRC there's Cahn-Hilliard codes that end up with `SEVENTH` order quadrature to deal with weighted mass matrices with third-order bases.

Just a thought, though.  It's really unlikely to have a mesh that's messed up badly enough to have a singular element but not badly enough for fourth-order quadrature to catch it.  But it's also unlikely to have a mesh that's messed up in a way that an interior rule won't catch it but a side rule will, and you're (smartly, IMHO) also testing on sides to catch those.

roystgnr replied:
Oh!  And I'd pull this (and the FE, and the side FE, and the side quadrature rule) out of the element loop.  We can save some memory reallocation and often skip some reinitialization that way.

GiudGiud replied:
fifth is fine

GiudGiud replied:
I moved the quadrature and the FE type out of the loop. I guess this is not the part that matters

GiudGiud replied:
ok it looks good actually now

GiudGiud committed:
Address Roy's review
- adapt language for adaptivity types
- give more information about the problematic elements, not just the centroid
- allow the user to tune the number of locartions output
- avoid false negatives on NODELEM

Co-authored-by: roystgnr <roy@stogners.org>

GiudGiud commented on framework/doc/content/source/meshgenerators/MeshDiagnosticsGenerator.md:
```diff
@@ -0,0 +1,136 @@
+Turn this check on with the [!param](/Mesh/MeshDiagnosticsGenerator/examine_element_volumes) parameter.
+
+A minimum and a maximum element volume can be specified to the `MeshDiagnosticsGenerator` using
+the [!param](/Mesh/MeshDiagnosticsGenerator/minimum_element_volumes) and the
```
does not trip element volume check!
however it trips:
```
*** Warning ***
The following warning occurred in the object "diag", of type "MeshDiagnosticsGenerator".

Number of elements overlapping (centroid-based heuristics): 14
```
and 
```

*** Warning ***
The following warning occurred in the object "diag", of type "MeshDiagnosticsGenerator".

Number of elements with a negative Jacobian: 16
```


is that ok?
do the NODELEM has negative Jacobian? There's 16 of them
Is it ok to report them as overlapping?

GiudGiud replied:
oh actually the negative Jacobian is reporting this
```
Element type not supported:NODEELEM
```
so it s being miscounted. will fix

roystgnr replied:
They definitely count as overlapping; that's totally fair to report.

I'm surprised by the negative Jacobian results, though; let me see if I can reproduce that in gdb and figure out what's wrong.  I bet it's throwing a *different* error, e.g. something about how we can't return a gradient in 0-D.  If that's the case then one of my other recommendations (prerequest only `get_JxW()`) would fix it.

GiudGiud replied:
yeah it s a different errot it's "not supported"

probably because of lagrange fe

GiudGiud replied:
handled better now. Reports 0 negative Jacobians

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+    {
+      if (elem->volume() <= _min_volume)
+      {
+        if (num_tiny_elems < 10)
```
done, tweakable from input

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+      if (elem->volume() <= _min_volume)
+      {
+        if (num_tiny_elems < 10)
+          _console << "Element too small detected with centroid : " << elem->true_centroid()
```
changed. get_info will give all one could ever want tbh, not sure why I went for position.

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+  // This deliberately does not trust the mesh to know whether it's already prepared or not
+  mesh->prepare_for_use();
+
+  if (_check_sidesets_orientation != "NO_CHECK")
```
sounds good on it

GiudGiud replied:
done

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+          continue;
+
+      // Form a parent, of the same type as the elements we are trying to combine!
+      std::unique_ptr<Elem> parent;
+      if (elem_type == QUAD4 || elem_type == QUAD8 || elem_type == QUAD9)
+        parent = std::make_unique<Quad4>();
+      else if (elem_type == HEX8 || elem_type == HEX20 || elem_type == HEX27)
+        parent = std::make_unique<Hex8>();
+      else if (elem_type == TRI3 || elem_type == TRI6 || elem_type == TRI7)
+        parent = std::make_unique<Tri3>();
+      else if (elem_type == TET4 || elem_type == TET10 || elem_type == TET14)
+        parent = std::make_unique<Tet4>();
+      else
+        continue;
```
was looking for that one smh did not find it

GiudGiud replied:
done. Hits an assert when trying to refine high order tet and tri

roystgnr replied:
Could you give me a stack trace?  I've spent the day (and returned to it in the night...) coding in a panic because you made me realize we don't have enough unit test coverage on TRI7 or TET14 refinement... but TRI6 and TET10 refinement should have been in dbg in CI for a decade, and the new tests I'm adding for the others aren't showing issues.

roystgnr replied:
Actually, looking through it after some sleep, we do have TRI7 and TET14 refinement in CI too.  The new unit tests are more focused and added better coverage and caught some bugs in not-yet-tested infinite element methods though, so my panic wasn't a complete waste.  It still leaves me baffled as to what assert you're triggering and I'm failing to, is all.

GiudGiud replied:
Keep in mind that I am not setting all the nodes to these second order elements. That s probably the problem

roystgnr replied:
Ah!  Yeah, that would definitely be an issue.  Weird that it wasn't a problem with tensor product elements too, though.

Would `Elem::build(Elem::first_order_equivalent_type(elem_type))` fix it, then?

GiudGiud replied:
for sure as that reverts to the previous thing I was doing just a lot shorter in terms of code

GiudGiud replied:
just did that

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+        std::set<const Elem *> all_elements;
+        elem->find_point_neighbors(*interior_node, all_elements);
+
+        if (elem_type == QUAD4 || elem_type == QUAD8 || elem_type == QUAD9)
```
I m planning to do that when I work on mesh coarsening already. e.g. move everything needed to MeshCoarseningUtils and meshUtils, as it makes sense

GiudGiud committed:
Split the main routine into separate subroutines for each check

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+            break;
+          }
+        }
+        if (!found)
```
ok added another heuristic

GiudGiud committed:
Avoid re-creating quadrature and fe_type all the time

GiudGiud committed:
Roy's review:
Use the elem type of the children to build the parents for detecting adaptivity non-conformality
Dont clone the mesh for every node that may host AMR adaptivity
Avoid re-initializing the FE all the time when check jacobian signs

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+      // We have to make a copy because adding the new parent element to the mesh
+      // will modify the mesh for the analysis of the next nodes
+      // Make a copy of the mesh, add this element
+      auto mesh_copy = mesh->clone();
+      auto parent_ptr = mesh_copy->add_elem(parent.release());
```
it works nice nice

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+      for (auto & elem : elements)
+      {
+        // If the node is not part of this element's nodes, it is a
+        // case of non-conformality
```
added in the docs

GiudGiud commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1196 @@
+          const auto coarse_side = coarse_elem->side_ptr(coarse_side_i);
+
+          // We did not find the side of the coarse neighbor near the refined elements
+          // It could be that it's not planar and the fine elements nodes do not lie on it
```
true true.

GiudGiud replied:
removed that one sentence

GiudGiud commented on framework/doc/content/source/meshgenerators/MeshDiagnosticsGenerator.md:
```diff
@@ -0,0 +1,136 @@
+
+Turn this check on with the [!param](/Mesh/MeshDiagnosticsGenerator/search_for_adaptivity_nonconformality) parameter.
+
+In the special case of meshes where uniform h-refinement was performed, the `MeshDiagnosticsGenerator` can detect
```
Yeah I meant uniform as in every fine element of a coarse parent is the same size. Which boils down to isotropic in an element

GiudGiud commented:
should be all addressed

GiudGiud commented:
ok i ll fix this tonight maybe

YaqiWang commented on framework/src/meshgenerators/MeshDiagnosticsGenerator.C:
```diff
@@ -0,0 +1,1262 @@
+  // Options for the output level
+  MooseEnum chk_option("NO_CHECK INFO WARNING ERROR", "NO_CHECK");
+
+  params.addParam<MooseEnum>(
```
Another check we can add is to make sure a boundary side is covered by a side set. We used to have an issue with Cubit that meshes are not properly merged that ends up with some boundary side in the inside of the domain, which are typically not covered by any side set.

If some checks are more serious, I suggest to change the default for the parameter. For example, that `examine_element_overlap` could be pretty bad, we might want to make it default to warning.

GiudGiud replied:
it s situation dependent. overlap is totally fine for nodeelem elements for example.

GiudGiud replied:
the external boundary totally covered is a good idea. Will have to be in the next iteration of adding diagnostics

GiudGiud replied:
for cubit internal sides, we could add also a "make_sure_these_sidesets_are_external"

GiudGiud commented:
Test failures unrelated.

GiudGiud commented:
Thanks for the review!!
